pyhap/characteristic.py

```diff
@@ -189,50 +151,44 @@ def override_properties(self, properties=None, valid_values=None):
-        .. seealso:: accessory_driver.publish
-
-        :raise RuntimeError: When the broker is not set.
-        """
-        if not self.broker:
-            raise RuntimeError('Notify failed, because broker is not set')
-
-        data = {
-            'type_id': self.type_id,
-            'value': self.value,
+        """Notify clients about a value change. Sends the value."""
+        self.broker.publish(self.value, self)
+
+    def to_dict(self):
+        """Create a dict repr of this object for json serialization."""
+        json_dict = {
+            'iid': self.broker.iid_manager.get_iid(self),
+            'type': str(self.type_id).upper(),
+            'description': self.display_name,
+            'perms': self.properties['Permissions'],
+            'format': self.properties['Format'],
         }
-        self.broker.publish(data, self)
 
-    def to_HAP(self, iid_manager):
-        """Create a HAP representation of this Characteristic. Sends the value.
-
-        :param iid_manager: IID manager to query for this object's IID.
-        :type iid_manager: IIDManager
-
-        :return: A HAP representation.
-        :rtype: dict
-        """
-        hap_rep = {
-            "iid": iid_manager.get_iid(self),
-            "type": str(self.type_id).upper(),
-            "description": self.display_name,
-            "perms": self.properties["Permissions"],
-            "format": self.properties["Format"],
-        }
+        if self.properties['Format'] in HAP_FORMAT.NUMERIC:
+            json_dict.update({k: self.properties[k] for k in
+                              self.properties.keys() & _HAP_NUMERIC_FIELDS})
+        elif self.properties['Format'] == HAP_FORMAT.STRING:
+            if len(self.value) > 64:
+                json_dict['maxLen'] = min(len(self.value), 256)
+        if HAP_PERMISSIONS.READ in self.properties['Permissions']:
+            json_dict['value'] = self.value
 
-        if self.properties["Format"] in HAP_FORMAT.NUMERIC:
-            value_info = {k: self.properties[k] for k in
-                          self.properties.keys() & _HAP_NUMERIC_FIELDS}
-        else:
-            value_info = dict()
+        return json_dict
 
-        if self.properties["Format"] == HAP_FORMAT.STRING:
-            if len(self.value) > 64:
-                value_info["maxLen"] = min(len(self.value), 256)
-        if HAP_PERMISSIONS.READ in self.properties["Permissions"]:
-            value_info["value"] = self.value
+    def update_value(self, value):
+        """Called from broker for value change in Home app.
 
-        hap_rep.update(value_info)
-        return hap_rep
+        Change self.value to value and call callback.
+        The callback method must return a boolean that indicates if successful.
+        """
+        self.value = value
+        self.notify()
+        if self.setter_callback:
+            return self.setter_callback(value)
+        return True
```

Comment:
Rather than returning `True`/`False` I think the callback should raise an exception on failure. The exception could be different depending on what went wrong.

We can then map the exceptions to the HAP protocol status values.

Using `True`/`False` feels a little bit like err values in C.

Reply:
I would think that `True / False` should be enough for our proposes, but you are right in that exceptions would allow for more options. @ikalchev What do you think?

Reply:
I would point out that `True`/`False` is not backwards compatible, and requires that users remember to return a `True` value on success (forgetting to return a value would result in `None`). This does not strike me as very pythonic.

Reply:
I just looked at Apple's spec - there are 11 error codes (which I need to add btw). If we use exceptions, we will force the Driver from understanding these exceptions and mapping them to error codes - in this case, why not `set_value` to return one of the error codes instead? What is more, returning an error code would allow "internal" errors to more "voicefully" propagate with an exception.

Reply:
There might be 11 error codes, but only some would be valid for a callback i.e. `Request denied due to insufficient privileges.` `Unable to communicate with requested service, e.g. the power to the accessory was turned off.` `Resource is busy, try again.` `Accessory received an invalid value in a write request.`

The rest are errors that should be managed by py-HAP. i.e. `Operation timed out.` could be thrown by setting a timeout on the callback (either via threads or asyncio).

Reply:
True, but let's keep in mind that tomorrow Apple may add 10 more, 4 of which would make sense (for example).

Reply:
What do you guys think if I exclude the error handling here and just call the callback method. That way we can implement it in another PR together with the complete error handling.

Reply:
Sounds good :)

Reply:
@ikalchev still doesn't feel pythonic. We would then need a bunch of static vars somewhere to match the docs, or rely on users parsing the docs.

Not sure how best to move this convo into another PR.

Reply:
@thomaspurchas maybe continue here: #75

Reply:
Let's think on that, it's not such a big deal, whatever we chose.

BTW there is a discord chat, but I haven't mentioned that here because it would make things more hard to follow.

Reply:
I've copied a summary of this thread into there.